content/blog/2019-10-28-self/index.md

```diff
@@ -0,0 +1,176 @@
+object and points to the clone family for its prototype. If the instance ever
+overrides one of its methods, the creates a new clone family to preserve
+semantics and propagate behavior changes to all clones of a prototype.
+
+<img src="without-maps.png" alt="drawing" width="300"/>
+<img src="with-maps.png" alt="drawing" width="300"/>
+
+### Customized Compilation
+
+The rampant dynamism in Self programs forces compilers to conservatively
+generate slow code. For example `x < y` is a method call on the object `x` and
+requires knowing the precise type of `x` to generate optimized code for `<`.
+The paper mentions that contemporary Smalltalk-80 compilers restricted the
+customization of primitive methods and control structures to allow for
+specialized code generation. Instead of imposing these restrictions, the Self
+compiler will generate specialized code for every _receiver_ object at runtime.
+This means that at the first invocation of `<` on an integer, the compiler will
+specialize `<` with check to see if the type matches to an `integer` and
+generate a single compare instruction when this is case. When the type of
+receiver has not been encountered before, Self will default to a method send
+to the object.
+
+### Message Inlining and Splitting
+
+Inlining remains one of the most crucial optimization in modern compilers,
+enabling other optimizations to become more effective. However, inlining
+becomes impossible in presence of dynamic method calls if the type of the
+receiver object is unknown. This is an even bigger problem for Self since
+most of program is a sequence of dynamic calls.
+
+The Self compiler performs two optimizations, message _inlining_ and
+_splitting_ to enable efficient execution. Message _inlining_ works in a similar
+fashion to customized compilation -- If the type of a receiver is know, either
+at initial compilation through a dataflow analysis or at runtime, inline the
+method body at the call location. When type information for an object is lost
+due to control flow splits, _message splitting_ generates specialized code for
+possible receiver types and guards them using type tests. If the type of
+receiver matches, fast code can be executed.
+
+### Programming Environment Support
+
+The Self compiler support _incremental recompilation_ and _source-level
+debugging_ by keeping track of the provenances of various method
+specializations in a map.
+
+Incremental Recompilation occurs when the compiler observes a change in the
+programming environment and invalidates compiled methods associated with the
+affected data. At compile time, the compiler creates a dependency list encoding
+the information required for specialization. For example, if a method was
+inlined from a clone's parent, updating the parent pointer or the method body
+in the parent should invalidate the specialized code. Since the compiler
+selectively invalidates code, methods that weren't modified can still execute
+with specialized code.
```

Comment:
This is the weirdest part of Self, and the one that has not stuck with us today: the interactive, live-editable code environment. It's maybe worth pointing out that this is the most anachronistic part of the paper as a result. V8 doesn't have an interactive, persistent, first-class IDE, for example, so it doesn't have to worry about this as much.